fri/src/layers.rs

```diff
@@ -0,0 +1,378 @@
+use crate::folder::MultiplicativeFriFolder;
+use anyhow::Error;
+use ark_ff::PrimeField;
+use ark_poly::EvaluationDomain;
+use std::sync::Arc;
+
+#[allow(dead_code)]
+pub trait FriLayer<F: PrimeField, E: EvaluationDomain<F>> {
+    fn get_layer_size(&self) -> usize;
+    fn get_domain(&self) -> E;
+    fn get_layer(&self) -> Result<Vec<F>, Error>;
+    fn eval_at_points(&self, required_indices: &[usize]) -> Vec<F>;
+    fn clone_arc(&self) -> Arc<dyn FriLayer<F, E>>;
+}
+
+#[derive(Clone)]
+pub struct FriLayerReal<F: PrimeField, E: EvaluationDomain<F>> {
+    layer_size: usize,
+    domain: E,
+    evaluation: Vec<F>,
+}
+
+#[allow(dead_code)]
+impl<F: PrimeField, E: EvaluationDomain<F>> FriLayerReal<F, E> {
+    pub fn new(domain: E, evaluation: Vec<F>) -> Self {
+        Self {
+            layer_size: domain.size(),
+            domain,
+            evaluation,
+        }
+    }
+
+    pub fn new_from_prev_layer(prev_layer: &dyn FriLayer<F, E>) -> Self {
```

Comment:
This just clones the layer?

Reply:
It is used to create `FriLayerReal` from the previous `FriLayerProxy`, both the structs implement `FriLayer` trait.